
#
# Patch managed by http://www.holgerschurig.de/patcher.html
#

--- native/src/c++/common/base/util/StringBuffer.cpp~dangling-pointer
+++ native/src/c++/common/base/util/StringBuffer.cpp
@@ -231,7 +231,9 @@
         int ldiff = tlen - flen ;
 
         // reallocate if needed
+        size_t p_off = p - s;     // remember position in string
         getmem(length() + ldiff);
+        p = s + p_off;            // ensure that p is valid again
         // check is there is a remainder after the replaced token
         if( p[flen] ) {
             tail = new BCHAR[length()];
