
#
# Patch managed by http://www.holgerschurig.de/patcher.html
#

--- native/src/c++/common/vocl/VConverter.cpp~string_overlap
+++ native/src/c++/common/vocl/VConverter.cpp
@@ -128,7 +128,8 @@
                 wchar_t* paramName = new wchar_t[wcslen(token) + 1];
                 token[paramIndex - token] = '\0';
                 wcscpy(paramName, token);
-                wcscpy(token, ++paramIndex);
+                ++paramIndex;
+                memmove(token, paramIndex, (strlen(paramIndex) + 1) * sizeof(*paramIndex));
 
                 wchar_t* paramVal = new wchar_t[wcslen(token) + 1];
                 wcscpy(paramVal, token);
