
#
# Patch managed by http://www.holgerschurig.de/patcher.html
#

--- native/src/c++/common/spds/spdsutils.cpp~sync-mode-names
+++ native/src/c++/common/spds/spdsutils.cpp
@@ -38,9 +38,11 @@
         return SYNC_ONE_WAY_FROM_CLIENT;
     else if (bstrcmp(syncMode,T("refresh")) == 0)
         return SYNC_REFRESH_FROM_SERVER;
-    else if (bstrcmp(syncMode,T("refresh-server")) == 0)
+    else if (bstrcmp(syncMode,T("refresh-server")) == 0 ||
+             bstrcmp(syncMode,T("refresh-from-server")) == 0)
         return SYNC_REFRESH_FROM_SERVER;
-    else if (bstrcmp(syncMode,T("refresh-client")) == 0)
+    else if (bstrcmp(syncMode,T("refresh-client")) == 0 ||
+             bstrcmp(syncMode,T("refresh-from-client")) == 0)
         return SYNC_REFRESH_FROM_CLIENT;
     //--------- Funambol extension --------------------
     else if (bstrcmp(syncMode, T("addrchange")) == 0)
