
#
# Patch managed by http://www.holgerschurig.de/patcher.html
#

--- native/src/c++/common/vocl/VObject.cpp~vcard-30-lineend
+++ native/src/c++/common/vocl/VObject.cpp
@@ -122,10 +122,8 @@
     WString strVObject;
     const wchar_t* eof;
 
-    if(version && !wcscmp(version,TEXT("3.0")))
-        eof = TEXT("\n");
-    else
-        eof = TEXT("\r\n");
+    // vcard 2.1 and 3.0 both use \r\n as line ending
+    eof = TEXT("\r\n");
 
     for (int i=0; i<properties->size(); i++) {
         VProperty *property;
